File: ios/Classes/FlutterAvoInspectorMethods.swift
//
//  FlutterAvoInspectorMethods.swift
//  flutter_avo_inspector
//
//  Created by Chandra Abdul Fattah on 02/02/23.
//

import Foundation
import AvoInspector
import IosAnalyticsDebugger

class FlutterAvoInspectorMethods {
    static func trackEventParams(
        avo: AvoInspector,
        debugger: AnalyticsDebugger,
        call: FlutterMethodCall,
        result: @escaping FlutterResult
    ){
        let args = call.arguments as! [String : Any]
        
        let name = args[Arguments.eventName] as! String
        let params = args[Arguments.eventParams] as? [String: Any]
        
        let track = avo.trackSchema(fromEvent: name, eventParams: params ?? [:] )
        
        var debuggerProps: [DebuggerProp] = []
        let debuggerErrors: [DebuggerPropError] = []
        for(key, value) in params ?? [:] {
            debuggerProps.append(DebuggerProp(
                id: key, withName: key, withValue: String(describing: value)
            ))
        }
        debugger.publishEvent(name, withTimestamp: NSNumber(value: NSDate().timeIntervalSince1970), withProperties: debuggerProps, withErrors: debuggerErrors)
        
        result(Constants.successResult)
    }

    static func logging(call: FlutterMethodCall, result: @escaping FlutterResult){
        let args = call.arguments as! [String : Any]

        let isEnable = args[Arguments.isEnableLogging] as! Bool
        AvoInspector.setLogging(isEnable)
        result(Constants.successResult)
    }

    static func isLogging(result: @escaping FlutterResult){
        let isEnableLogging = AvoInspector.isLogging()
        result(isEnableLogging)
    }

    static func setBatchSize(call: FlutterMethodCall, result: @escaping FlutterResult){
        let args = call.arguments as! [String : Any]

        let batchSize = args[Arguments.batchSize] as! Int32

        AvoInspector.setBatchSize(batchSize)
        result(Constants.successResult)
    }

    static func batchSize(call: FlutterMethodCall, result: @escaping FlutterResult){
        let value = AvoInspector.getBatchSize()
        result(value)
    }

    static func setBatchFlushSeconds(call: FlutterMethodCall, result: @escaping FlutterResult){
        let args = call.arguments as! [String : Any]
        let batchFlushSeconds = args[Arguments.batchFlushSeconds] as! Int32
        AvoInspector.setBatchFlushSeconds(batchFlushSeconds)
        result(Constants.successResult)
    }

    static func batchFlushSeconds(call: FlutterMethodCall, result: @escaping FlutterResult){
        let batchFlushSeconds = AvoInspector.getBatchFlushSeconds()
        result(batchFlushSeconds)
    }
    
    static func showVisualInspector(debugger: AnalyticsDebugger, call: FlutterMethodCall, result: @escaping FlutterResult){
        let args = call.arguments as! [String : Any]

        let mode = args[Arguments.visualInspectorMode] as! Int
        var viMode: AvoVisualInspectorType = AvoVisualInspectorType.Bubble
        
        if(mode == 0){
            viMode = AvoVisualInspectorType.Bar
        }
        
        if(viMode == AvoVisualInspectorType.Bubble){
            debugger.showBubble()
        }
        else {
            debugger.showBarDebugger()
        }
        result(Constants.successResult)
    }
    
    static func hideVisualInspector(debugger: AnalyticsDebugger, call: FlutterMethodCall, result: @escaping FlutterResult){
        debugger.hide()
        result(Constants.successResult)
    }
}

File: android/src/main/kotlin/com/solusibejo/flutter_avo_inspector/FlutterAvoInspectorMethods.kt
package com.solusibejo.flutter_avo_inspector

import android.app.Activity
import app.avo.inspector.AvoInspector
import app.avo.inspector.VisualInspectorMode
import com.solusibejo.flutter_avo_inspector.consts.Arguments
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

object FlutterAvoInspectorMethods {
    fun trackEventParams(
        avo: AvoInspector,
        call: MethodCall,
        result: MethodChannel.Result
    ){
        val name = call.argument<String>(Arguments.eventName)!!
        val params = call.argument<Map<String, Any>?>(Arguments.eventParams)

        avo.trackSchemaFromEvent(
            name,
            params ?: emptyMap<String, Any>()
        )

        result.success(200)
    }

    fun trackEventJson(
        avo: AvoInspector,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        val name = call.argument<String>(Arguments.eventName)!!
        val jsonObject = call.argument<JSONObject?>(Arguments.eventJson)

        val track = avo.trackSchemaFromEvent(name, jsonObject)

        if(track.entries.isNotEmpty()){
            result.success(200)
        }
        else {
            result.error("500", "Track using params JSONObject failed",
                "Entries is empty, track using params JSONObject failed")
        }
    }

    fun logging(
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        val isEnable = call.argument<Boolean>(Arguments.isEnableLogging)!!
        AvoInspector.enableLogging(isEnable)
        result.success(200)
    }

    fun isLogging(
        result: MethodChannel.Result
    ) {
        val isEnableLogging = AvoInspector.isLogging()
        result.success(isEnableLogging)
    }

    fun setBatchSize(
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        val batchSize = call.argument<Int>(Arguments.batchSize)!!
        AvoInspector.setBatchSize(batchSize)
        result.success(200)
    }

    fun batchSize(
        result: MethodChannel.Result
    ) {
        val size = AvoInspector.getBatchSize()
        result.success(size)
    }

    fun setBatchFlushSeconds(
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        val batchFlushSeconds = call.argument<Int>(Arguments.batchFlushSeconds)!!
        AvoInspector.setBatchFlushSeconds(batchFlushSeconds)
        result.success(200)
    }

    fun batchFlushSeconds(
        result: MethodChannel.Result
    ) {
        val batchFlushSeconds = AvoInspector.getBatchFlushSeconds()
        result.success(batchFlushSeconds)
    }

    fun showVisualInspector(
        activity: Activity,
        avo: AvoInspector,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        val mode = call.argument<Int>(Arguments.visualInspectorMode)!!
        var viMode = VisualInspectorMode.BUBBLE

        if(mode == 0) viMode = VisualInspectorMode.BAR

        avo.showVisualInspector(activity, viMode)
        result.success(200)
    }

    fun hideVisualInspector(
        activity: Activity,
        avo: AvoInspector,
        result: MethodChannel.Result
    ) {
        avo.hideVisualInspector(activity)
        result.success(200)
    }
}